### PR TITLE
Errors take 3

### PR DESCRIPTION
**Pull request by imply-cheddar**
This is the 3rd take at creating a unified Exception in the DruidException.  Given that I had input on the previous two attempts, I figured I'd take the next stab at offering something that matches the input that I had provided.

`DruidException` and `ErrorResponse` have class level javadoc discussing how it's all setup and why, so I would recommend starting there.  Then branch out and look at some of the points where DruidExceptions are created (DruidPlanner and various of the other places).

There is a compatibility layer in this PR that tries to continue mimicing the old way that QueryException was getting serialized out, but the current PR treats that as best-effort.  Specifically, anything that started its life as a DruidException will be an "unknown" error type when serialized back as a QueryException.  The error message should make it through, but the type characterization will be different.  This should only happen in the middle of roll-forward/roll-back processes, so it will be short-lived and, given that it preserves the message, is hopefully a decent compromise.

Additionally, as I was adjusting tests, I got tired of the SQL tests taking roughly 5 minutes to run through the entire suite, so I looked into why it was doing that and through moving some initialization around and fixing some time-based async tests to be based on latches instead, was able to get that timing down to ~1.5 minutes, so some of those changes are mixed in here as well.

#### Release note

A new unified exception for surfacing errors to users has been introduced.  It is partially compatible with the old way of reporting error messages, response codes remain the same, all fields that previously existed on the response will continue to exist and be populated, including the errorMessage.  But some error messages have changed to be more consumable by humans and some cases will have the message restructured.  Any clients that are actively looking at the error message to make decisions on what to do should explore the changes to the error messages that they care about, but if all a client looks at is the response code, there should be no impact.  

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/error/DruidException.java:370]: nit: javadocs

**Review by clintropolis — COMMENTED** [review]
the design of `DruidException` as an internal holder for error stuff seems like a nice improvement over the current `QueryException` stuff, and splitting the serde out of it should allow stuff to handle things in a much more flexible manner depending on the channel the exception is going to be pushed through (nice friendly messages for users and different stuff for logs, etc).

while looking over stuff i got a bit distracted with individual error messages, so feel free to ignore most of those comments since the things largely aren't new error messages, just minor adjustments to existing things

**Review by abhishekagarwal87 — COMMENTED** [review]
some commentary that I had forgotten to publish. I think steering developers to think about persona is a good step in the right direction. I also think that similarly, we should be able to have developers also write an action associated with the error message. For example, if router can't find any brokers, we log  
```
org.apache.druid.java.util.common.ISE: No default server found!`
```
This doesn't really offer much help to a new user. The error itself could have been better rewritten. But it would also be nice if it says, go and check your broker's health.

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/error/DruidException.java:250]: It might be "file a bug" or it might just be "this error message is completely impenetrable, we know that, but don't have the time to make it so that we can make a useful error message, so we just mark it as intended for the DEVELOPER audience and move on".  It would be great if there are no errors for this persona (or if they are basically all just DEFENSIVE exceptions)...

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/error/DruidException.java:36]: Generally speaking, IAE/ISE would disappear and all be replaced by `DruidException`.  We can/definitely would continue to have rules that map IAE/ISE into `DruidException` (probably as a DEVELOPER-focused persona exception), but in the fullness of time, we'd probably want IAE/ISE to be replaced with `DruidException` instances.

File: processing/src/test/java/org/apache/druid/matchers/DMatchers.java
Comment (imply-cheddar) [reply processing/src/test/java/org/apache/druid/matchers/DMatchers.java:26]: Going away from the PR and coming back to it, I have no clue what it is either :P  Will look into it and update.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidRexExecutor.java
Comment (imply-cheddar) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidRexExecutor.java:134]: I took it as the expression has a return type and that return type isn't matching the input of another expression or something like that.  Which is why it had the suggestion to cast...  But, I was mostly just keeping what was in existence around rather than creating anything new.

File: processing/src/main/java/org/apache/druid/error/InvalidSqlInput.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/error/InvalidSqlInput.java:22]: Yes, it is weird that `sql` is not in `processing` (see what I did there? :P)

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/error/DruidException.java:294]: so lets say someone issues a `UNION` query with de-duplicating while only supports `UNION ALL`. would you consider that as a `501` or `400`? Such planning errors are very common. I am concerned about the error code for such scenarios because a cluster admin might have set an alert on 5xx but not on 4xx. Since the former will indicate the server being bad while the latter indicates the user being bad.

**Comment by imply-cheddar** [comment]
@abhishekagarwal87 

On your suggestion about encouraging an action be included.  You gave a great example that explains what this PR is all about.  Specifically, think of this error message

```
org.apache.druid.java.util.common.ISE: No default server found!
```

If someone said that it was "Persona.USER", I would hope that the reviewer would be like, "uhhh, why will a user know what a default server is?  Your message is not well aligned with your persona" and ask for the message to be updated.

If I were to adjust that exception to this new model and not change the message, I would likely make it `Persona.DEVELOPER` because it's so obtuse, only a developer would be able to actually make sense of what it means.  Of course, the exception is actually happening at a point where there is some behavior that is meaningful for the end user.

We could adjust the message to be an ADMIN persona message and say

```
Router unable to find a broker, check that brokers are up and active
```

Or a USER persona message and say

```
Cannot find a queryable server, contact your cluster administrator to validate that all services are operational
```

Both of these include an action, like you suggest, but that action becomes apparent and clear once we actually identify the persona.  My *hope* is that by identifying the persona, reviewers will start to validate that the exception messages align with the intended persona and we all police each other for creating good error messages.  I.e. I'm hoping that just identifying the persona is enough to help us do better.

Alternatively, for USER-facing messages, I could add the requirement for yet-another field which is "potential actions" (or something similar) which should be populated with a message describing actions that can be taken.  I'd personally like to keep it to just a message and code-review based policing.  But I'm also willing to add the field if I'm in the minority...

Tagging other people whose name also appears in the conversation part of this PR to see if there are opinions one way or the other: @gianm @kfaraz @clintropolis

**Comment by abhishekagarwal87** [comment]
Now I am also wondering how we will pick a persona :) For example, in this case, I would like to tell the user that you really need to go and check the logs of the broker service. The action of checking error logs is something that an admin persona can do. But most people when they are setting up druid for the first time, are end-users as well as an admin. The error occurs directly in response to something that an end-user is doing. 

I can return a user-persona-facing error saying that "Contact your cluster administrator to make sure that services are operational" but it's difficult for new users to figure out where to go from there. 

So, in this case, maybe we return an error message meant for the admin persona even though this error is in user-initiated request-response path. We leave it to cluster admins if they want to filter messages that match admin persona, in case they are not exposing druid directly. For a new user setting up druid for first time, however they will get all these messages unfiltered. If you have something similar in mind, we should ideally document that developers should avoid adding potentially sensitive information in messages that are meant for user persona.

**Comment by imply-cheddar** [comment]
@abhishekagarwal87 once again, you've provided a great example!  Essentially, you went through the thought process that I'm hoping we can just do as a part of every exception message that we generate.

In this case, making it ADMIN-focused and having more details about server names and stuff makes a lot of sense to me.  And the argument that someone who is trying to "hide" Druid should be able to siphon off this message makes a lot of sense to me too.  So, I think that I'm now convinced that the message in question *should* be targeting the ADMIN persona and specifically mention that you should check that the brokers are up and running.

To me, this type of conversation, back-and-forth and your eventual conclusion is the whole goal of the adjustments to the error messages.  And, once we change that exception to be ADMIN focused, the code will include the fact that this is an ADMIN-focused message.  So any future person who comes across it can clearly see the intentions of who gets the message and why, if they feel the message needs updating, they can have their own internal dialog and motivate their feelings by either saying it should target a different persona OR it needs more details to target the current persona.

So anyway, I think that yes, you are thinking along the same lines that I'm thinking at this point.  I'm all for adding documentation about whatever, but I'm unclear on where you think it would be added for the most impact.  Ultimately, this has to become part of the culture of the project rather than be a pure documentation thing, but I can document it anywhere you think it can help ;)

**Comment by abhishekagarwal87** [comment]
Re culture, I agree. I think we add these bits in the APIs javadocs, that should be more than enough. And then reviewers can just say, "The error message isn't so great. please read the javadocs for reference"

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/error/DruidException.java:294]: TBH, I'm fine with both.  The 501 code is actually just a hold-over from the QueryException rather than something I'm actually introducing.  If anything, reviewing this change has caused it to be more clear that we are returning 501s in some instances, leading to a question of "WTF", why?  Which I think is a totally valid question ;).

Fwiw, I have a TODO in there because this is something I wanted to revisit myself.  At this point, the only usages of this in the code in this PR are

https://github.com/apache/druid/pull/14004/files#diff-483e089932b8faa5ca9669607eab7fcba883d27e2899a7c9fba2f37afb9de97bR599-R602

Which is what happens when we get a query planning failure and no message at all.  I don't know if/when this actually happens, so I'm unclear on whether it's actually meaningful.  That said, with the switch to the `DruidException`, we should be able to eliminate the whole `plannerContext().getPlanningError()` thing that is the source of potentially not having a message (and also is the source of red-herring error messages more often than not), which should, in turn, eliminate the need for this category.  Regardless, it needs to exist to maintain what QueryException is doing...

**Comment by imply-cheddar** [comment]
Okay, had a clean bill. Aside from some static checks.  Going to push those up.  @abhishekagarwal87 @clintropolis I'd love to get this merged today or tomorrow if possible, please do what you need to do to feel comfortable approving.

image

File: processing/src/main/java/org/apache/druid/error/DruidException.java
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/error/DruidException.java:36]: i think there is still a use case for them maybe for low level exceptions that don't have the full picture because they are missing some context to provide a meaningful exception at the surface. In these cases I expect the code area which does have the context will catch IAE/ISE/UOE etc and then extract the details and decorate with additional context such as column and table name, etc.

That seems preferable to me than having to push all of this information down all of the way to the bottom, particularly in more 'hot' areas of the code since passing arguments isn't completely free. Though, I suspect some of these cases are probably better to use more specific exceptions dedicated to a purpose, _or_ I guess we could still throw `DruidException` but catch it at a higher level to decorate with the additional context, so I think we'll just have to see how it works out in practice.

File: processing/src/main/java/org/apache/druid/error/ErrorResponse.java
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/error/ErrorResponse.java:149]: there is a similar existing method, `Evals.asString` though its primarily used for processing data for expressions and query time stuff, so i don't feel super strongly about using it here or not

File: processing/src/main/java/org/apache/druid/error/InvalidSqlInput.java
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/error/InvalidSqlInput.java:22]: heh, that isn't really possible without also merging druid-server into processing, or alternatively splitting druid-sql operators and stuff into processing and putting the schema and higher level stuff that builds on broker stuff in druid-server (which might also be nice, but would make testing slightly more painful maybe?)

File: processing/src/main/java/org/apache/druid/error/README.md
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/error/README.md:21]: is this still true? (maybe it should be refreshed to align with the current code before we merge?)

also any possible way to add something like a 'tl;dr' to summarize the important parts? 😛

File: processing/src/test/java/org/apache/druid/error/ErrorResponseTest.java
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/error/ErrorResponseTest.java:32]: since ErrorResponse is an 'over the wire' thingo, should serde tests be here? (or did i miss them somewhere?)

File: server/src/test/java/org/apache/druid/server/QueryResourceTest.java
Comment (clintropolis) [inline server/src/test/java/org/apache/druid/server/QueryResourceTest.java:357]: ah i guess this is the serde test i was looking for?

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:325]: any reason not to just make all of these `instanceof` checks, especially since some of the catch still have `instanceof`?

**Review by clintropolis — APPROVED** [review]
a few minor questions, but I'll go ahead and throw a +1 on this since it seems an overall improvement :metal:

**Comment by clintropolis** [comment]
also, should there be any web-console changes to go with the adjustments to error response?

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java
Comment (cheddar) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:325]: not really, I think I started doing that and then stopped for some reason, I don't remember why though.

File: extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/tuple/sql/ArrayOfDoublesSketchSqlAggregatorTest.java
Comment (abhishekagarwal87) [reply extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/tuple/sql/ArrayOfDoublesSketchSqlAggregatorTest.java:440]: ok. I think that this got replaced with "Query planning failed for unknown reason, our best guess is this"

**Review by abhishekagarwal87 — APPROVED** [review]
+1 since I don't have major blocker comments and this PR is prone to conflicts. @clintropolis does have a point around console changes. Ideally, there shouldn't be any because if there are, we likely broke compatibility somewhere. Or console is doing things it shouldn't do.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java
Comment (cheddar) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1627]: I'm anti that messaging because it pretends like the following error message has a high likelihood of being the cause.  My experience has told me that this is rarely the case and most likely this error message means "something happened, I'm gonna tell you something that is just as likely to be wrong as it is right, good luck".  So I don't want to pretend like it's giving people a good answer when it generally is not.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java
Comment (imply-cheddar) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1662]: Technically various of the SQL exceptions get caught and used to tell the volcano planner that it generated a bad query and to keep looking.  None of those are converted to DruidException yet.  Also, most of the places that were throwing `QueryException` are still throwing `QueryException` because we need a release that has the `DruidException` in it before we can force those forward.

So... really, any place that potentially has other uses just didn't get converted.  We will need to go back and review them with the understanding that we now have an exception whose sole purposes is to be delivered back to the user and decide if the current handling is good.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQInsertTest.java
Comment (imply-cheddar) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQInsertTest.java:1079]: Yes, it could.  I'm not sure which, but you are either seeing stuff left behind when I first started converting tests and hadn't yet build the Matcher to make it simpler (and I missed it in going back to simplify). OR, with fixing the conflicts after 2 months dormant, I will admit that I didn't have all of the context still in my head and perhaps did something a bit rote.  Either way, it's much simpler to switch to the `invalidSqlContains()`.

**Comment by cheddar** [comment]
The last run of the tests was near complete (there processing tests for jdk11 and jdk17 still running, everything else, including ITs had passed), so I went ahead and merged instead of risk another set of conflicts cropping up and causing yet another CI run.

**Pull request merged**
